slime/backends/fsdp_utils/actor.py

```diff
@@ -165,60 +176,73 @@ def compute_log_prob(
-            max_len = max(len(t) for t in batch_tokens)
-            padded_tokens = [t + [self.tokenizer.pad_token_id] * (max_len - len(t)) for t in batch_tokens]
-            padded_loss_masks = [
-                # -1 because its the loss mask for logprob
-                [0] * (len(t) - len(l) - 1) + l + [0] * (max_len - len(t))
-                for l, t in zip(batch_loss_masks, batch_tokens)
-            ]
-            batch = {
-                "tokens": torch.tensor(padded_tokens, dtype=torch.long, device=torch.cuda.current_device()),
-                "loss_masks": torch.tensor(padded_loss_masks, dtype=torch.int, device=torch.cuda.current_device()),
-                "rewards": torch.tensor(
-                    rollout_data["rewards"][i : i + self.args.micro_batch_size],
-                    dtype=torch.float,
-                    device=torch.cuda.current_device(),
-                ),
-                "raw_reward": rollout_data["raw_reward"][i : i + self.args.micro_batch_size],
-            }
 
-            if self.args.multimodal_keys:
-                processed_media = {}
-                for sample_prompt in batch_prompts:
-                    for media_part in sample_prompt:
-                        media_type = media_part.get("type")
-
-                        if media_type == "image":
-                            path = media_part.get("path")
-                            if path:
-                                if "pixel_values" not in processed_media:
-                                    processed_media["pixel_values"] = []
-                                image = Image.open(path).convert("RGB")
-                                inputs = self.vlm_processor(images=image, return_tensors="pt")
-                                processed_media["pixel_values"].append(inputs.pixel_values)
-
-                # Stack and move all processed media to the GPU for the batch
-                for key, tensor_list in processed_media.items():
-                    if tensor_list:
-                        batch[key] = torch.cat(tensor_list).to(
-                            device=torch.cuda.current_device(), dtype=torch.bfloat16
-                        )
-
-            padded_batches.append(batch)
-        return padded_batches
-
-    def train(self, rollout_id, rollout_data_ref):  # type: ignore[override]
+        packed_batches = []
+        mbs_size_list = []
+        dp_size = dist.get_world_size()
+        local_batch_size = self.args.global_batch_size // dp_size
+
+        assert (
+            self.args.global_batch_size % dp_size == 0
+        ), f"global_batch_size {self.args.global_batch_size} is not divisible by dp_world_size {dp_size}"
+        # Use global_batch_size for splitting when max_tokens_per_gpu is enabled
+        if (
+            hasattr(self.args, "max_tokens_per_gpu")
+            and self.args.max_tokens_per_gpu is not None
+            and self.args.use_dynamic_batch_size
+        ):
```

Comment:
we can assume the args has `args.use_dynamic_batch_size` and if `args.use_dynamic_batch_size` is set the `args.max_tokens_per_gpu is not None`.